src/v/coproc/materialized_topics_frontend.cc

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "coproc/materialized_topics_frontend.h"
+
+#include "cluster/metadata_cache.h"
+#include "cluster/topics_frontend.h"
+#include "coproc/logger.h"
+#include "vlog.h"
+namespace coproc {
+
+materialized_topics_frontend::materialized_topics_frontend(
+  ss::sharded<cluster::topics_frontend>& topics_frontend) noexcept
+  : _topics_frontend(topics_frontend) {}
+
+void materialized_topics_frontend::topic_creation_resolved(
+  std::vector<cluster::topic_result> results) {
+    for (const auto& result : results) {
+        auto found = _topics.find(result.tp_ns);
+        vassert(
+          found != _topics.end(), "Missing promise with associated event");
+        if (result.ec == cluster::errc::success) {
+            vlog(
+              coproclog.info, "Materialized topic created: {}", result.tp_ns);
+            found->second->set_value();
+        } else if (result.ec == cluster::errc::topic_already_exists) {
+            vlog(
+              coproclog.info,
+              "Materialzed log has come into existance via "
+              "another node: {}",
+              result.tp_ns);
+            found->second->set_value();
+        } else {
+            found->second->set_exception(
+              materialized_topic_replication_exception(fmt::format(
+                "Failed to replicate materialized topic: {}, error: {}",
+                result.tp_ns,
+                result.ec)));
+        }
+        _topics.erase(found);
+    }
+}
+
+ss::future<> materialized_topics_frontend::create_materialized_topics(
+  std::vector<cluster::non_replicable_topic> topics,
+  model::timeout_clock::time_point timeout) {
+    std::vector<ss::future<>> all;
+    std::vector<cluster::non_replicable_topic> todos;
+    for (auto& topic : topics) {
+        auto [itr, success] = _topics.insert(
+          {topic.name, ss::make_lw_shared<ss::shared_promise<>>()});
+        if (!success) {
+            all.emplace_back(itr->second->get_shared_future());
+        } else {
+            todos.emplace_back(std::move(topic));
+        }
+    }
+    if (!todos.empty()) {
+        auto f = _topics_frontend.local()
+                   .create_non_replicable_topics(std::move(todos), timeout)
```

Comment:
we should handle exceptions here, if this would throw, we would never make promises available